src/main/resources/db/migration/V1__tables.sql

```diff
@@ -6,52 +6,58 @@ DROP TABLE IF EXISTS versions CASCADE;
 
 CREATE TABLE objects
 (
-    id         BIGSERIAL PRIMARY KEY,
-    hash       CHAR(64) NOT NULL,
-    url        TEXT     NOT NULL,
-    client_id  TEXT     NOT NULL,
-    content    BYTEA    NOT NULL,
-    is_deleted BOOLEAN  NOT NULL DEFAULT FALSE
+    id         BIGINT  GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+    hash       TEXT    NOT NULL,
+    url        TEXT    NOT NULL,
+    client_id  TEXT    NOT NULL,
+    content    BYTEA   NOT NULL,
+    is_deleted BOOLEAN NOT NULL DEFAULT FALSE,
+    CONSTRAINT hash_is_unique UNIQUE (hash)
 );
 
+CREATE UNIQUE INDEX idx_objects_url ON objects (url) WHERE NOT is_deleted;
+CREATE INDEX idx_objects_client_id ON objects (client_id);
+
 CREATE TABLE object_log
 (
-    id            BIGSERIAL PRIMARY KEY,
-    operation     CHAR(3) NOT NULL,
+    id            BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+    operation     TEXT   NOT NULL,
     new_object_id BIGINT,
     old_object_id BIGINT,
-    CHECK (operation IN ('INS', 'UPD', 'DEL')),
-    CHECK (
-            operation = 'INS' AND old_object_id IS NULL AND new_object_id IS NOT NULL OR
-            operation = 'UPD' AND new_object_id IS NOT NULL AND new_object_id IS NOT NULL OR
-            operation = 'DEL' AND old_object_id IS NOT NULL AND new_object_id IS NULL
-        )
+    CHECK (CASE operation
+           WHEN 'INS' THEN old_object_id IS NULL AND new_object_id IS NOT NULL
+           WHEN 'UPD' THEN new_object_id IS NOT NULL AND new_object_id IS NOT NULL
+           WHEN 'DEL' THEN old_object_id IS NOT NULL AND new_object_id IS NULL
+           ELSE FALSE
+           END
+        ),
+    FOREIGN KEY (new_object_id) REFERENCES objects (id) ON DELETE RESTRICT ON UPDATE RESTRICT,
+    FOREIGN KEY (old_object_id) REFERENCES objects (id) ON DELETE RESTRICT ON UPDATE RESTRICT
 );
 
+CREATE INDEX idx_object_log_new_object_id ON object_log (new_object_id) WHERE new_object_id IS NOT NULL;
```

Comment:
Doesn't PG create an index for FKs automatically?

Reply:
No, an index is not required on a foreign key column, just highly recommended. See https://www.cybertec-postgresql.com/en/index-your-foreign-key/ for some background information and when such an index is not necessary (rarely).

Reply:
cool!